### DISCUSSION TITLE
How to merge a default Storybook config with a user override which both contain async `webpackFinal` fields

### DISCUSSION THREAD

**Discussion by virtuoushub** [storybookjs/storybook#18150] [Help] [answered]
This is slightly Storybook adjacent, but since it has to do with Storybook's [`webpackFinal` field](https://storybook.js.org/docs/react/configure/webpack#extending-storybooks-webpack-config), I figure I'll ask here.

Are there recommended patterns for merging async webpack configs?

I am currently using a library, `webpack-merge` which has this known limitation:

> Note that Promises are not supported! If you want to return a configuration wrapped within a Promise, merge inside one. Example: Promise.resolve(merge({ ... }, { ... })).

https://github.com/survivejs/webpack-merge#limitations

What I am trying to do is provide a set of baked in defaults for Storybook's webpack config, and then allow a user to extend/override those defaults to meet their needs.

**Comment by shilman** [comment, marked as answer]
The way Storybook presets work is that they accumulate. So assume you have two addons with presets:

### Addon1
```js
export const webpackFinal: async (config, options) => merge(config, { some: options })
```

### Addon2
```js
export const webpackFinal: async (config, options) => merge(config, { other: options })
```

### main.js
```
module.exports = {
  addons: ['Addon1', 'Addon2']
}
```

Then I believe the `config` passed to `Addon1` will be Storybook's default, the `config` passed to `Addon2` will contain `Addon1`'s changes. Those `webpackFinal` functions are async so they can await whatever you need to populate the values, and the presets will handle the coordination.

Hope that helps!
